File: Java/src/Medium/Permutations.java
// https://leetcode.com/problems/permutations/?envType=problem-list-v2&envId=backtracking
package Medium;

import java.util.List;
import java.util.ArrayList;

public class Permutations{

    List<List<Integer>> res = new ArrayList<>();
 
    public List<List<Integer>> permute(int[] nums){
        
        List<Integer> nums1 = new ArrayList<>();
        for(int num: nums){
            nums1.add(num);
        }
        backtrack(nums1, new ArrayList<>());
        return res;

    }

    /* Does not work */
    // private void backTrack(List<Integer> nums, List<Integer> path){

    //     if(nums.isEmpty()){
    //         res.add(new ArrayList<>(path));
    //         System.out.println(res);
    //         return;
    //     }

    //     for(int i=0; i < nums.size(); i++){
    //         List<Integer> subList = new ArrayList<>(nums.subList(0, i));
    //         subList.addAll(new ArrayList<>(nums.subList(i+1, nums.size())));
    //         System.out.println(subList);
    //         path = new ArrayList<>(path);
    //         path.add(nums.get(i));
    //         System.out.println(path);
    //         backTrack(subList, path);
    //     }
    // }

    private void backtrack(List<Integer> nums, List<Integer> path) {
        if (nums.isEmpty()) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < nums.size(); i++) {
            List<Integer> remaining = new ArrayList<>(nums);
            List<Integer> newPath = new ArrayList<>(path);
            newPath.add(remaining.remove(i));
            backtrack(remaining, newPath);
        }
    }

    public static void main(String[] args) {
        Permutations perm = new Permutations();
        System.out.println(perm.permute(new int[]{1, 2, 3}));
        System.out.println(perm.permute(new int[]{0}));
    }

}

File: Python/Medium/Permutations.py
# https://leetcode.com/problems/permutations/?envType=problem-list-v2&envId=backtracking

def permute(nums: list[int]):

    """ Youtube """
    # def backtrack(nums, path):
    #     if not nums:
    #         result.append(path)
    #         return
    #     for i in range(len(nums)):
    #         backtrack(nums[:i] + nums[i+1:], path + [nums[i]])
    
    # result = []
    # backtrack(nums, [])
    # return result

    """ Neetcode """
    if len(nums) == 0:
        return [[]]

    perms = permute(nums[1:])
    res = []

    for p in perms:
        for i in range(len(p) + 1):
            p_copy = p[:]
            p_copy.insert(i, nums[0])
            res.append(p_copy)
    
    return res

if __name__ == '__main__':

    n = list(map(int, input().split()))

    print(permute(n))